// ngx-joyride/esm2015/lib/components/button/button.component.js
import { Component, Input, EventEmitter, Output } from "@angular/core";
export class JoyrideButtonComponent {
    constructor() {
        this.clicked = new EventEmitter();
    }
    onClick() {
        this.clicked.emit();
    }
}
JoyrideButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'joyride-button',
                template: "<button (mouseleave)=\"hover=false\" (mouseover)=\"hover=true\"\r\n    [ngStyle]=\"{'background-color': hover ? '#fff' : color, \r\n                'color': hover ? color : '#fff',\r\n                'border-color' : hover ? color : 'transparent'}\"\r\n    class=\"joyride-button\" (click)=\"onClick()\">\r\n    <ng-content></ng-content>\r\n</button>",
                styles: [".joyride-button{text-transform:uppercase;border:2px solid transparent;outline:none;padding:6px 12px;font-size:12px;font-weight:700;color:#fff;background-color:#3b5560;cursor:pointer}.joyride-button:hover{color:#3b5560;border:2px solid #3b5560;background-color:#fff}"]
            },] }
];
JoyrideButtonComponent.propDecorators = {
    color: [{ type: Input }],
    clicked: [{ type: Output }]
};
