// angular2_photoswipe/esm2015/lib/lightbox-adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @return {?}
 */
export function ANGULAR2_PHOTO_SWIPE_FACTORY() {
    return new DefaultLightboxAdapter();
}
/**
 * @abstract
 */
export class LightboxAdapter {
}
LightboxAdapter.decorators = [
    { type: Injectable, args: [{ providedIn: 'root', useFactory: ANGULAR2_PHOTO_SWIPE_FACTORY },] }
];
/** @nocollapse */ LightboxAdapter.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: ANGULAR2_PHOTO_SWIPE_FACTORY, token: LightboxAdapter, providedIn: "root" });
if (false) {
    /** @type {?} */
    LightboxAdapter.prototype.getThumbBoundsFn;
    /** @type {?} */
    LightboxAdapter.prototype.getDoubleTapZoom;
    /** @type {?} */
    LightboxAdapter.prototype.isClickableElement;
    /** @type {?} */
    LightboxAdapter.prototype.addCaptionHTMLFn;
    /** @type {?} */
    LightboxAdapter.prototype.index;
    /** @type {?} */
    LightboxAdapter.prototype.showHideOpacity;
    /** @type {?} */
    LightboxAdapter.prototype.showAnimationDuration;
    /** @type {?} */
    LightboxAdapter.prototype.hideAnimationDuration;
    /** @type {?} */
    LightboxAdapter.prototype.bgOpacity;
    /** @type {?} */
    LightboxAdapter.prototype.spacing;
    /** @type {?} */
    LightboxAdapter.prototype.allowPanToNext;
    /** @type {?} */
    LightboxAdapter.prototype.maxSpreadZoom;
    /** @type {?} */
    LightboxAdapter.prototype.loop;
    /** @type {?} */
    LightboxAdapter.prototype.pinchToClose;
    /** @type {?} */
    LightboxAdapter.prototype.closeOnScroll;
    /** @type {?} */
    LightboxAdapter.prototype.closeOnVerticalDrag;
    /** @type {?} */
    LightboxAdapter.prototype.mouseUsed;
    /** @type {?} */
    LightboxAdapter.prototype.escKey;
    /** @type {?} */
    LightboxAdapter.prototype.arrowKeys;
    /** @type {?} */
    LightboxAdapter.prototype.history;
    /** @type {?} */
    LightboxAdapter.prototype.galleryUID;
    /** @type {?} */
    LightboxAdapter.prototype.galleryPIDs;
    /** @type {?} */
    LightboxAdapter.prototype.errorMsg;
    /** @type {?} */
    LightboxAdapter.prototype.barsSize;
    /** @type {?} */
    LightboxAdapter.prototype.timeToIdle;
    /** @type {?} */
    LightboxAdapter.prototype.timeToIdleOutside;
    /** @type {?} */
    LightboxAdapter.prototype.loadingIndicatorDelay;
    /** @type {?} */
    LightboxAdapter.prototype.closeEl;
    /** @type {?} */
    LightboxAdapter.prototype.captionEl;
    /** @type {?} */
    LightboxAdapter.prototype.fullscreenEl;
    /** @type {?} */
    LightboxAdapter.prototype.zoomEl;
    /** @type {?} */
    LightboxAdapter.prototype.shareEl;
    /** @type {?} */
    LightboxAdapter.prototype.counterEl;
    /** @type {?} */
    LightboxAdapter.prototype.arrowEl;
    /** @type {?} */
    LightboxAdapter.prototype.preloaderEl;
    /** @type {?} */
    LightboxAdapter.prototype.tapToClose;
    /** @type {?} */
    LightboxAdapter.prototype.tapToToggleControls;
    /** @type {?} */
    LightboxAdapter.prototype.clickToCloseNonZoomable;
    /** @type {?} */
    LightboxAdapter.prototype.closeElClasses;
    /** @type {?} */
    LightboxAdapter.prototype.indexIndicatorSep;
    /** @type {?} */
    LightboxAdapter.prototype.shareButtons;
    /** @type {?} */
    LightboxAdapter.prototype.enableBootstrap4;
}
export class DefaultLightboxAdapter extends LightboxAdapter {
    constructor() {
        super(...arguments);
        this.allowPanToNext = true;
        this.spacing = 0.12;
        this.bgOpacity = 1.0;
        this.mouseUsed = false;
        this.loop = true;
        this.pinchToClose = true;
        this.closeOnScroll = true;
        this.closeOnVerticalDrag = true;
        this.hideAnimationDuration = 333;
        this.showAnimationDuration = 333;
        this.showHideOpacity = false;
        this.escKey = true;
        this.arrowKeys = true;
    }
}
DefaultLightboxAdapter.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    DefaultLightboxAdapter.prototype.allowPanToNext;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.spacing;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.bgOpacity;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.mouseUsed;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.loop;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.pinchToClose;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.closeOnScroll;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.closeOnVerticalDrag;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.hideAnimationDuration;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.showAnimationDuration;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.showHideOpacity;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.escKey;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.arrowKeys;
    /** @type {?} */
    DefaultLightboxAdapter.prototype.enableBootstrap4;
}
