// angular2_photoswipe/esm5/lib/gallery/gallery.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList, ViewChild, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import * as PhotoSwipe from 'photoswipe';
import * as PhotoSwipeUI_Default from 'photoswipe/dist/photoswipe-ui-default';
import { GalleryItemComponent } from '../gallery-item/gallery-item.component';
import { NgpService } from '../ngp.service';
import { LightboxAdapter } from '../lightbox-adapter';
var GalleryComponent = /** @class */ (function () {
    function GalleryComponent(ngp, adapter) {
        this.ngp = ngp;
        this.adapter = adapter;
        this.shareLinkClick = new EventEmitter();
        this.subscriptions = [];
        this.images = [];
        this.isBootstrapEnabled = this.adapter.enableBootstrap4;
    }
    /**
     * @return {?}
     */
    GalleryComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this_1 = this;
        this.images = (/** @type {?} */ (this.galleryItems.toArray().map((/**
         * @param {?} cp
         * @return {?}
         */
        function (cp) {
            // listen for clicks;
            _this_1.subscriptions.push(cp.clicked.subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return _this_1.onClick(data); })));
            return cp.image;
        }))));
    };
    /**
     * @return {?}
     */
    GalleryComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        function (sub) { return sub.unsubscribe(); }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    GalleryComponent.prototype.onClick = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.openPhotoSwipe(data, this.galleryElement);
    };
    /**
     * @private
     * @param {?} img
     * @param {?} galleryDOM
     * @return {?}
     */
    GalleryComponent.prototype.openPhotoSwipe = /**
     * @private
     * @param {?} img
     * @param {?} galleryDOM
     * @return {?}
     */
    function (img, galleryDOM) {
        this.adapter.galleryUID = galleryDOM.nativeElement.getAttribute('data-pswp-uid');
        this.adapter.index = img.id;
        /** @type {?} */
        var PSWP = (/** @type {?} */ (this.ngp.LightboxElement.nativeElement));
        this.pswp = new PhotoSwipe(PSWP, PhotoSwipeUI_Default, this.getImagesAsPhotoswipe(), this.adapter);
        /** @type {?} */
        var _this = this;
        this.pswp.listen('shareLinkClick', (/**
         * @param {?} e
         * @param {?} target
         * @return {?}
         */
        function (e, target) {
            _this.shareLinkClick.emit({ e: e, target: target });
        }));
        this.pswp.init();
        return false;
    };
    /**
     * @private
     * @return {?}
     */
    GalleryComponent.prototype.getImagesAsPhotoswipe = /**
     * @private
     * @return {?}
     */
    function () {
        return this.images.map((/**
         * @param {?} image
         * @return {?}
         */
        function (image) {
            return {
                src: image.largeUrl,
                w: image.width,
                h: image.height,
                pid: image.id,
                title: image.description,
                author: image.author
            };
        }));
    };
    GalleryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngp-gallery',
                    template: "<div class=\"angular2_photoswipe\" [ngClass]=\"{'card-columns': isBootstrapEnabled }\"  #ngpGallery id=\"sampleId\" itemscope itemtype=\"http://schema.org/ImageGallery\">\n  <ng-content select=\"ngp-gallery-item\"></ng-content>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    GalleryComponent.ctorParameters = function () { return [
        { type: NgpService },
        { type: LightboxAdapter }
    ]; };
    GalleryComponent.propDecorators = {
        galleryElement: [{ type: ViewChild, args: ['ngpGallery', { static: true },] }],
        galleryItems: [{ type: ContentChildren, args: [GalleryItemComponent,] }],
        galleryId: [{ type: Input }],
        shareLinkClick: [{ type: Output }]
    };
    return GalleryComponent;
}());
export { GalleryComponent };
if (false) {
    /** @type {?} */
    GalleryComponent.prototype.galleryElement;
    /** @type {?} */
    GalleryComponent.prototype.galleryItems;
    /** @type {?} */
    GalleryComponent.prototype.galleryId;
    /** @type {?} */
    GalleryComponent.prototype.shareLinkClick;
    /** @type {?} */
    GalleryComponent.prototype.subscriptions;
    /** @type {?} */
    GalleryComponent.prototype.isBootstrapEnabled;
    /** @type {?} */
    GalleryComponent.prototype.pswp;
    /** @type {?} */
    GalleryComponent.prototype.images;
    /**
     * @type {?}
     * @private
     */
    GalleryComponent.prototype.ngp;
    /**
     * @type {?}
     * @private
     */
    GalleryComponent.prototype.adapter;
}
