// ngx-editor/esm2015/lib/modules/menu/simple-command/simple-command.component.js
import { Component, HostBinding, Input } from '@angular/core';
import { SimpleCommands } from '../MenuCommands';
import Icon from '../../../icons';
import { NgxEditorService } from '../../../editor.service';
import { SharedService } from '../../../services/shared/shared.service';
export class SimpleCommandComponent {
    constructor(ngxeService, sharedService) {
        this.ngxeService = ngxeService;
        this.sharedService = sharedService;
        this.isActive = false;
        this.disabled = false;
        this.update = (view) => {
            const { state } = view;
            const command = SimpleCommands.get(this.name);
            this.isActive = command.isActive(state);
            this.disabled = !command.canExecute(state);
        };
        this.editorView = this.sharedService.view;
        this.pluginUpdateSubscription = this.sharedService.plugin.update.subscribe((view) => {
            this.update(view);
        });
    }
    toggle(e) {
        e.preventDefault();
        if (e.button !== 0) {
            return;
        }
        const { state, dispatch } = this.editorView;
        const command = SimpleCommands.get(this.name);
        command.execute(state, dispatch);
    }
    getTitle(name) {
        return this.ngxeService.locals.get(name);
    }
    ngOnInit() {
        this.html = Icon.get(this.name);
    }
    ngOnDestroy() {
        this.pluginUpdateSubscription.unsubscribe();
    }
}
SimpleCommandComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-simple-command',
                template: "<div class=\"NgxEditor__MenuItem--IconContainer\" [innerHTML]=\"html | sanitizeHtml\" (mousedown)=\"toggle($event)\"\n  [title]=\"getTitle(name)\">\n</div>\n",
                styles: [""]
            },] }
];
SimpleCommandComponent.ctorParameters = () => [
    { type: NgxEditorService },
    { type: SharedService }
];
SimpleCommandComponent.propDecorators = {
    name: [{ type: Input }],
    isActive: [{ type: HostBinding, args: ['class.NgxEditor__MenuItem--Active',] }],
    disabled: [{ type: HostBinding, args: ['class.NgxEditor--Disabled',] }]
};
