// ngx-joyride/esm2015/lib/components/arrow/arrow.component.js
import { Component, Input, ViewEncapsulation } from "@angular/core";
export class JoyrideArrowComponent {
    constructor() {
        this.position = 'top';
    }
}
JoyrideArrowComponent.decorators = [
    { type: Component, args: [{
                selector: 'joyride-arrow',
                template: "<div [class.joyride-arrow__top]=\"position == 'top'\"\r\n     [class.joyride-arrow__bottom]=\"position == 'bottom'\"\r\n     [class.joyride-arrow__left]=\"position == 'left'\"\r\n     [class.joyride-arrow__right]=\"position == 'right'\">\r\n</div>",
                encapsulation: ViewEncapsulation.None,
                styles: [".joyride-arrow__top{border-bottom:11px solid #fff}.joyride-arrow__bottom,.joyride-arrow__top{border-left:11px solid transparent;border-right:11px solid transparent}.joyride-arrow__bottom{border-top:11px solid #fff}.joyride-arrow__right{border-left:11px solid #fff}.joyride-arrow__left,.joyride-arrow__right{border-bottom:11px solid transparent;border-top:11px solid transparent}.joyride-arrow__left{border-right:11px solid #fff}"]
            },] }
];
JoyrideArrowComponent.propDecorators = {
    position: [{ type: Input }]
};
