// ngx-editor/esm2015/lib/commands/Link.js
import { toggleMark } from 'prosemirror-commands';
import { isMarkActive } from 'ngx-editor/helpers';
import { removeLink } from 'ngx-editor/commands';
const defaultOptions = {
    strict: true
};
class Link {
    update(attrs = {}, state, dispatch) {
        const { schema } = state;
        const type = schema.marks.link;
        if (!type) {
            return false;
        }
        return toggleMark(type, attrs)(state, dispatch);
    }
    insert(text, attrs = {}, state, dispatch) {
        const { schema, tr } = state;
        const type = schema.marks.link;
        if (!type) {
            return false;
        }
        const node = schema.text(text, [schema.marks.link.create(attrs)]);
        dispatch(tr.replaceSelectionWith(node, false));
        return true;
    }
    isActive(state, options = defaultOptions) {
        if (options.strict) {
            return true;
        }
        const { schema } = state;
        return isMarkActive(state, schema.marks.link);
    }
    remove(state, dispatch) {
        return removeLink()(state, dispatch);
    }
    canExecute(state) {
        return this.update({}, state, null);
    }
}
export default Link;
