// ngx-editor/esm2015/lib/commands/ListItemToggle.js
import { liftListItem, wrapInList } from 'prosemirror-schema-list';
import { isNodeActive } from 'ngx-editor/helpers';
class ListItemToggle {
    constructor(isBulletList = false) {
        this.isBulletList = false;
        this.isBulletList = isBulletList;
    }
    getType(schema) {
        return this.isBulletList ? schema.nodes.bullet_list : schema.nodes.ordered_list;
    }
    execute(state, dispatch) {
        const { schema } = state;
        const type = this.getType(schema);
        if (!type) {
            return false;
        }
        if (this.isActive(state)) {
            return liftListItem(schema.nodes.list_item)(state, dispatch);
        }
        return wrapInList(type)(state, dispatch);
    }
    isActive(state) {
        const { schema } = state;
        const type = this.getType(schema);
        if (!type) {
            return false;
        }
        return isNodeActive(state, type);
    }
    canExecute(state) {
        return this.execute(state, null);
    }
}
export default ListItemToggle;
