// angular2_photoswipe/esm2015/lib/gallery/gallery.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList, ViewChild, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import * as PhotoSwipe from 'photoswipe';
import * as PhotoSwipeUI_Default from 'photoswipe/dist/photoswipe-ui-default';
import { GalleryItemComponent } from '../gallery-item/gallery-item.component';
import { NgpService } from '../ngp.service';
import { LightboxAdapter } from '../lightbox-adapter';
export class GalleryComponent {
    /**
     * @param {?} ngp
     * @param {?} adapter
     */
    constructor(ngp, adapter) {
        this.ngp = ngp;
        this.adapter = adapter;
        this.shareLinkClick = new EventEmitter();
        this.subscriptions = [];
        this.images = [];
        this.isBootstrapEnabled = this.adapter.enableBootstrap4;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.images = (/** @type {?} */ (this.galleryItems.toArray().map((/**
         * @param {?} cp
         * @return {?}
         */
        cp => {
            // listen for clicks;
            this.subscriptions.push(cp.clicked.subscribe((/**
             * @param {?} data
             * @return {?}
             */
            (data) => this.onClick(data))));
            return cp.image;
        }))));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        sub => sub.unsubscribe()));
    }
    /**
     * @param {?} data
     * @return {?}
     */
    onClick(data) {
        this.openPhotoSwipe(data, this.galleryElement);
    }
    /**
     * @private
     * @param {?} img
     * @param {?} galleryDOM
     * @return {?}
     */
    openPhotoSwipe(img, galleryDOM) {
        this.adapter.galleryUID = galleryDOM.nativeElement.getAttribute('data-pswp-uid');
        this.adapter.index = img.id;
        /** @type {?} */
        const PSWP = (/** @type {?} */ (this.ngp.LightboxElement.nativeElement));
        this.pswp = new PhotoSwipe(PSWP, PhotoSwipeUI_Default, this.getImagesAsPhotoswipe(), this.adapter);
        /** @type {?} */
        const _this = this;
        this.pswp.listen('shareLinkClick', (/**
         * @param {?} e
         * @param {?} target
         * @return {?}
         */
        function (e, target) {
            _this.shareLinkClick.emit({ e, target });
        }));
        this.pswp.init();
        return false;
    }
    /**
     * @private
     * @return {?}
     */
    getImagesAsPhotoswipe() {
        return this.images.map((/**
         * @param {?} image
         * @return {?}
         */
        image => {
            return {
                src: image.largeUrl,
                w: image.width,
                h: image.height,
                pid: image.id,
                title: image.description,
                author: image.author
            };
        }));
    }
}
GalleryComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngp-gallery',
                template: "<div class=\"angular2_photoswipe\" [ngClass]=\"{'card-columns': isBootstrapEnabled }\"  #ngpGallery id=\"sampleId\" itemscope itemtype=\"http://schema.org/ImageGallery\">\n  <ng-content select=\"ngp-gallery-item\"></ng-content>\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
GalleryComponent.ctorParameters = () => [
    { type: NgpService },
    { type: LightboxAdapter }
];
GalleryComponent.propDecorators = {
    galleryElement: [{ type: ViewChild, args: ['ngpGallery', { static: true },] }],
    galleryItems: [{ type: ContentChildren, args: [GalleryItemComponent,] }],
    galleryId: [{ type: Input }],
    shareLinkClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    GalleryComponent.prototype.galleryElement;
    /** @type {?} */
    GalleryComponent.prototype.galleryItems;
    /** @type {?} */
    GalleryComponent.prototype.galleryId;
    /** @type {?} */
    GalleryComponent.prototype.shareLinkClick;
    /** @type {?} */
    GalleryComponent.prototype.subscriptions;
    /** @type {?} */
    GalleryComponent.prototype.isBootstrapEnabled;
    /** @type {?} */
    GalleryComponent.prototype.pswp;
    /** @type {?} */
    GalleryComponent.prototype.images;
    /**
     * @type {?}
     * @private
     */
    GalleryComponent.prototype.ngp;
    /**
     * @type {?}
     * @private
     */
    GalleryComponent.prototype.adapter;
}
