// ngx-editor/esm2015/lib/services/shared/shared.service.js
var _view;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class SharedService {
    constructor() {
        _view.set(this, void 0);
        this.customMenuRefChange = new Subject();
        this.plugin = {
            update: new Subject(),
            destroy: new Subject()
        };
    }
    set view(v) {
        __classPrivateFieldSet(this, _view, v);
    }
    get view() {
        return __classPrivateFieldGet(this, _view);
    }
    setCustomMenuRef(c) {
        this.customMenuRefChange.next(c);
    }
}
_view = new WeakMap();
SharedService.ɵprov = i0.ɵɵdefineInjectable({ factory: function SharedService_Factory() { return new SharedService(); }, token: SharedService, providedIn: "root" });
SharedService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
SharedService.ctorParameters = () => [];
