// angular2_photoswipe/esm2015/lib/gallery-item/gallery-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding } from '@angular/core';
import { Image } from '../image';
import { LightboxAdapter } from '../lightbox-adapter';
export class GalleryItemComponent {
    /**
     * @param {?} adapter
     */
    constructor(adapter) {
        this.adapter = adapter;
        this.clicked = new EventEmitter();
        this.isBootstrapEnabled = false;
        this.isBootstrapEnabled = this.adapter.enableBootstrap4;
    }
    /**
     * @return {?}
     */
    onClick() {
        this.clicked.emit(this.image);
        return false;
    }
}
GalleryItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngp-gallery-item',
                template: "<a href=\"{{image.largeUrl}}\" itemprop=\"contentUrl\" [attr.data-size]=\"image.width + 'x' + image.height\"\n    [attr.data-author]=\"image.author\" (click)=\"onClick()\">\n    <img src=\"{{image.thumbUrl}}\" [ngClass]=\"{'card-img-top': isBootstrapEnabled }\"  itemprop=\"thumbnail\" alt=\"{{image.description}}\" />\n</a>",
                styles: [""]
            }] }
];
/** @nocollapse */
GalleryItemComponent.ctorParameters = () => [
    { type: LightboxAdapter }
];
GalleryItemComponent.propDecorators = {
    image: [{ type: Input }],
    clicked: [{ type: Output }],
    isBootstrapEnabled: [{ type: HostBinding, args: ['class.card',] }]
};
if (false) {
    /** @type {?} */
    GalleryItemComponent.prototype.image;
    /** @type {?} */
    GalleryItemComponent.prototype.clicked;
    /** @type {?} */
    GalleryItemComponent.prototype.isBootstrapEnabled;
    /**
     * @type {?}
     * @private
     */
    GalleryItemComponent.prototype.adapter;
}
